src/client/src/producer.rs

```diff
@@ -96,82 +195,219 @@ impl TopicProducer {
-    partition_request.records.batches.push(batch);
-    topic_request.name = replica.topic.to_owned();
-    topic_request.partitions.push(partition_request);
+        let grouped = group_by_spu("TOPIC", &partitions, records_by_partition)
+            .await
+            .unwrap();
 
-    request.acks = 1;
-    request.timeout_ms = 1500;
-    request.topics.push(topic_request);
+        // SPU 0 should have partitions 0 and 1, but not 2
+        let spu_0 = grouped.get(&0).unwrap();
+        let partition_0 = spu_0.get(&0).unwrap();
 
-    trace!("produce request: {:#?}", request);
+        assert!(
+            partition_0
+                .iter()
+                .any(|record| record.value.as_ref() == b"A"),
+            "SPU 0/Partition 0 should contain record A",
+        );
+        assert!(
+            partition_0
+                .iter()
+                .any(|record| record.value.as_ref() == b"D"),
+            "SPU 0/Partition 0 should contain record D",
+        );
 
-    let response = leader.send_receive(request).await?;
+        let partition_1 = spu_0.get(&1).unwrap();
+        assert!(
+            partition_1
+                .iter()
+                .any(|record| record.value.as_ref() == b"B"),
+            "SPU 0/Partition 1 should contain record B",
+        );
+        assert!(
+            partition_1
+                .iter()
+                .any(|record| record.value.as_ref() == b"E"),
+            "SPU 0/Partition 1 should contain record E",
+        );
 
-    trace!("received response: {:?}", response);
+        assert!(
+            spu_0.get(&2).is_none(),
+            "SPU 0 should not contain Partition 2",
+        );
 
-    // process response
-    match response.find_partition_response(&replica.topic, replica.partition) {
-        Some(partition_response) => {
-            if partition_response.error_code.is_error() {
-                return Err(IoError::new(
-                    ErrorKind::Other,
-                    partition_response.error_code.to_sentence(),
-                )
-                .into());
-            }
-            Ok(())
-        }
-        None => Err(IoError::new(ErrorKind::Other, "unknown error").into()),
+        let spu_1 = grouped.get(&1).unwrap();
+        let partition_2 = spu_1.get(&2).unwrap();
+
+        assert!(
+            spu_1.get(&0).is_none(),
+            "SPU 1 should not contain Partition 0",
+        );
+
+        assert!(
+            spu_1.get(&1).is_none(),
+            "SPU 1 should not contain Partition 1",
+        );
+
+        assert!(
+            partition_2
+                .iter()
+                .any(|record| record.value.as_ref() == b"C"),
+            "SPU 1/Partition 2 should contain record C",
+        );
+
+        assert!(
+            partition_2
+                .iter()
+                .any(|record| record.value.as_ref() == b"F"),
+            "SPU 1/Partition 2 should contain record F",
+        );
+
+        Ok(())
     }
```

Comment:
need test with replication

Reply:
I think replication tests are more fitting as an integration test or e2e test. Let's make another issue for it